### PR TITLE
Closes #389: TaskSetup, TaskTeardown functionality

### LINKED ISSUE #389: Enhancement: TaskSetup and TaskTeardown
I think it would be useful to have actions that could be executed before and after each task, particularly if those blocks receive an instance of the current task (or metadata about the task).

For example, that would make it easy to report progress to TeamCity, or maybe add blocks around each task in TeamCity's output to make its build logs more readable.

### PR DESCRIPTION
**Pull request by jrnail23**
Initial design approach for #389, submitted for feedback. Code is not tested or documented.

I'm not in love with the procedural logic in `CakeEngine` -- I think my implementation could benefit from a bit of internal refactoring of that class's existing code.

### CODE REVIEW

File: src/Cake.Core/ITaskTeardownContext.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+using System;
+
+namespace Cake.Core
+{
+    public interface ITaskTeardownContext
```
Comment (patriksvensson) [inline src/Cake.Core/ITaskTeardownContext.cs]: One type per class.

File: src/Cake.Core/ITaskSetupContext.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Cake.Core
+{
+    public interface ITaskSetupContext
```
Comment (patriksvensson) [inline src/Cake.Core/ITaskSetupContext.cs]: One type per file.

File: src/Cake.Core/CakeEngine.cs [hunk 3]
```diff
@@ -219,12 +234,41 @@ private void ExecuteTask(ICakeContext context, IExecutionStrategy strategy, Stop
...
+                PerformTaskTeardown(context, strategy, task, stopWatch.Elapsed, false);
             }
 
             // Add the task results to the report.
             report.Add(task.Name, stopWatch.Elapsed);
         }
 
+        private void PerformTaskSetup(ICakeContext context, IExecutionStrategy strategy, ICakeTaskInfo task, bool skipped)
+        {
+            // Trying to stay consistent with the behavior of script-level Setup & Teardown (if setup fails, don't run the task, but still run the teardown)
```
Comment (patriksvensson) [inline src/Cake.Core/CakeEngine.cs]: I agree with this :smile:

File: src/Cake.Core/ITaskTeardownContext.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    public sealed class TaskTeardownContext : ITaskTeardownContext
+    {
+        public TaskTeardownContext(ICakeTaskInfo task, TimeSpan duration, bool skipped)
+        {
+            Task = task;
+            Duration = duration;
+            Skipped = skipped;
+        }
+
+        public ICakeTaskInfo Task { get; private set; }
```
Comment (patriksvensson) [inline src/Cake.Core/ITaskTeardownContext.cs]: Use readonly backing fields that you assign in the constructor and provide only getters.

**Comment by patriksvensson** [comment]
@jrnail23 Looks good to me! Left some comments about minor code standard stuff but other than that it looks good!

There are no tests or documentation for this, but you already stated that in your PR description.

Good work! :+1:

File: src/Cake.Core/ITaskTeardownContext.cs (hunk 4 above)
Comment (jrnail23) [inline src/Cake.Core/ITaskTeardownContext.cs]: :+1: this is actually also my preferred personal coding style.
How do you guys feel about null checks / guard clauses for the args on these kinds of things (that's another thing I typically do)?

Comment (patriksvensson) [inline src/Cake.Core/ITaskTeardownContext.cs]: Yes. Null checks are nice to have. I really hate `NullReferenceException` :wink:

**Comment by jrnail23** [comment]
@patriksvensson, everything is tested and documented now, ready to go at your convenience.

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 5]
```diff
@@ -629,6 +629,261 @@ public void Should_Execute_Error_Handler_Even_If_Exception_Was_Thrown_In_Error_R
...
+                engine.RegisterTask("A").Does(() => result.Add("Executing A"));
+                engine.RegisterTask("B").Does(() => result.Add("Executing B")).IsDependentOn("A");
+
+                // When
+                Record.Exception(() => engine.RunTarget(fixture.Context, fixture.ExecutionStrategy, "B"));
+
+                // Then
+                Assert.Equal(new List<string>(), result);
+                
+                //Assert.False(runTask, "Task A was executed although it shouldn't have been.");
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: This should be removed?

Comment (jrnail23) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: no, just the commented part.  I must've overlooked that line.

**Comment by jrnail23** [comment]
@patriksvensson, commented-out code has been removed

**Comment by jrnail23** [comment]
@patriksvensson, this should be ready to merge

**Comment by devlead** [comment]
@jrnail23 Merged, thanks! :+1: 
Just FYI `src/Cake.Core.Tests/Unit/CakeEngineTests.cs(880,21)` had an un used variable that XBuild didn't like, I've fixed that. rebased and took no credit ;)

**Pull request merged**